numpy/linalg/cholesky.cc

```diff
@@ -45,50 +30,87 @@ void Cholesky::make_args(int size) {
     for (int i = 0; i < n; i++) {
         r_mat[i * n + i] = 1;
     }
-    cblas_dgemm(CblasColMajor, CblasNoTrans, CblasTrans, n, n, n, 1.0, x_mat, n,
-                x_mat, n, n, r_mat, n);
+
+    static const char uplo = 'U';
+    static const char trans = 'T';
+    static const double alpha = 1.;
+    static const double beta = n;
+    dsyrk(&uplo, &trans, &n, &n, &alpha, x_mat, &n, &beta, r_mat, &n);
 
     // we now have r_mat = x_mat * x_mat' + n * np.eye(n)
     // copy back into x_mat
     memcpy(x_mat, r_mat, mat_size * sizeof(*x_mat));
 }
 
 void Cholesky::copy_args() {
-    memcpy(r_mat, x_mat, mat_size * sizeof(*x_mat));
+    // copy moved to compute()
 }
 
 void Cholesky::compute() {
+    // perform copy here.
+    static const int one = 1;
+    dcopy(&mat_size, x_mat, &one, r_mat, &one);
+
     // compute cholesky decomposition
     int info;
-    const char uplo = 'U';
+    static const char uplo = 'U';
     dpotrf(&uplo, &n, r_mat, &lda, &info);
     assert(info == 0);
 
     // we only want an upper triangular matrix
-    for (int i = 0; i < n - 1; i++) {
-        memset(&r_mat[i * n + i + 1], 0, (n - i - 1) * sizeof(*r_mat));
+    // in scipy, this is done in *potrf wrapper.
+    // https://github.com/scipy/scipy/blob/maintenance/1.3.x/scipy/linalg/flapack_pos_def.pyf.src#L85
+    for (int i = 0; i < n; i++) {
+        for (int j = i + 1; j < n; j++) {
+            r_mat[i * n + j] = 0.;
+        }
     }
 }
 
-bool Cholesky::test() {
+bool Cholesky::test(bool verbose) {
     clean_args();
     make_args(test_size);
-    memcpy(x_mat, x_mat_test, mat_size * sizeof(*x_mat));
     copy_args();
     compute();
 
-    return mat_equal(r_mat, r_mat_test, mat_size);
+    // verify that r_mat is upper triangular
+    for (int i = 0; i < n; i++) {
+        for (int j = i + 1; j < n; j++) {
+            if (r_mat[i * n + j] != 0.) {
+                if (verbose) {
+                    std::cerr << "r_mat is not upper triangular!" << std::endl;
+                }
+                return false;
+            }
+        }
+    }
+
+    // try to reconstruct x_mat from its Cholesky decomposition
+    static const double alpha = 1., beta = 0.;
```

Comment:
Because `X = L.T @ L`, we should be using `dsyrk` rather than dgemm.

Reply:
done!